security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,75 @@
+package pl.touk.nussknacker.ui.security.api
+
+import akka.http.scaladsl.server.Directive1
+import io.circe.parser._
+import io.circe.syntax._
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserData}
+import pl.touk.nussknacker.ui.security.accesslogic.{AnonymousAccess, ImpersonatedAccess}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticationManager(authenticationResources: AuthenticationResources) {
+
+  lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+  private val anonymousAccess    = new AnonymousAccess(authenticationResources)
+  private val impersonatedAccess = new ImpersonatedAccess(authenticationResources)
+
+  def authenticate(): Directive1[AuthenticatedUser] = {
+    anonymousAccess.anonymousUserRole match {
+      case Some(role) =>
+        authenticationResources.authenticate().optional.flatMap {
+          case Some(user) => impersonatedAccess.provideOrImpersonateUser(user)
+          case None =>
+            anonymousAccess.handleAuthorizationFailedRejection.tmap(_ =>
+              AuthenticatedUser.createAnonymousUser(Set(role))
+            )
+        }
+      case None => authenticationResources.authenticate().flatMap(impersonatedAccess.provideOrImpersonateUser)
```

Comment:
does it make sense to impersonate the anonymous user?

Reply:
Currently it would be possible if technical user sent impersonated user data with an `Anonymous` role in the header.
I don't think there is much need for impersonating anonymous user but since a user with `Impersonate` permission can impersonate any user he could also do that with an anonymous one.

Reply:
After the changes where we do not directly pass roles in the impersonation header and pass impersonated user's identity instead it does not seem to be possible anymore.